### PR TITLE
CHE-2344: Fixes repetitive saving snapshot of workspace

### LINKED ISSUE #2344: Snapshoting of machines failed if snapshot made second time for given machine
**Reproduction Steps:**
1. Create new machine.
2. Snapshot machine.
3. Stop machine.
4. Start machine from saved snapshot.
5. Attempt to snapshot machine.

**Expected behavior:**

Snapshot is made overwriting old one, possible display confirmation dialog to user.

**Observed behavior:**

Error shown:

```
2016-09-05 14:30:13,003[aceManager-3600]  [ERROR] [o.e.c.a.w.s.WorkspaceManager 745]    - Snapshot with id 'snapshot******' or combination of workspace id 'workspace*****' & environment 'default' & machine name  'default'  already exists
org.eclipse.che.api.machine.server.exception.SnapshotException: Snapshot with id 'snapshot*****' or combination of workspace id 'workspace******' & environment 'default' & machine name  'default'  already exists
    at com.codenvy.api.dao.mongo.SnapshotDaoImpl.saveSnapshot(SnapshotDaoImpl.java:92) ~[codenvy-hosted-platform-api-impl-4.7.0.jar:4.7.0]
    at org.eclipse.che.api.workspace.server.WorkspaceManager.createSnapshotSync(WorkspaceManager.java:740) [che-core-api-workspace-4.7.0.jar:4.7.0]
    at org.eclipse.che.api.workspace.server.WorkspaceManager.lambda$createSnapshot$0(WorkspaceManager.java:518) [che-core-api-workspace-4.7.0.jar:4.7.0]
    at org.eclipse.che.api.workspace.server.WorkspaceManager$$Lambda$238/772546380.run(Unknown Source) [che-core-api-workspace-4.7.0.jar:4.7.0]
    at org.eclipse.che.commons.lang.concurrent.CopyThreadLocalRunnable.run(CopyThreadLocalRunnable.java:28) [che-core-commons-lang-4.7.0.jar:4.7.0]
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_45]
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_45]
    at java.lang.Thread.run(Thread.java:745) [na:1.8.0_45]
```

**Che version:**    4.7
**OS and version:** Centos 7.2
**Docker version:** 1.11.1
**Che install:**  Codenvy 4.7

**Additional information:**
**_This problem is reproduced only hosted version CHE_**

### PR DESCRIPTION
**Pull request by garagatyi**
### What does this PR do?

Fixes repetitive saving snapshot of workspace.
Also does some cleanups.
### What issues does this PR fix or reference?

Fixes #2344 
### PR type
- [x] Minor change = no change to existing features or docs
### Minor change checklist
- [x] Tests provided / updated
- [x] Tests passed

### CODE REVIEW

**Comment by garagatyi** [comment]
@evoevodin @skabashnyuk Please review

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java [hunk 1]
```diff
@@ -757,6 +751,31 @@ boolean createSnapshotSync(WorkspaceRuntimeImpl runtime, String namespace, Strin
...
+
+                runtimes.removeSnapshot(oldSnapshot);
+            } catch (NotFoundException ignored) {
+                // Do nothing if no snapshot found
+            }
+
+            SnapshotImpl snapshot = runtimes.saveMachine(namespace,
+                                                         machine.getWorkspaceId(),
+                                                         machine.getId());
+            snapshotDao.saveSnapshot(snapshot);
```
Comment (voievodin) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java]: If this call fails, then docker image still exists, i think you should try to make a synthetic transaction and at least try to remove the image 

e.g.

```
try {
    runtimes.save
    dao.save
} catch {
    try {
         runtimes.remove
    } catch {
         // appropriate logging which says that DOCKER image wasn't removed and due to some error
    }
}
```

Comment (garagatyi) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java]: done

**Comment by voievodin** [comment]
ok

**Comment by skabashnyuk** [comment]
ок

**Pull request merged**